### PR TITLE
Add publish test results command 

### PR DESCRIPTION
**Pull request by AshwiniChalla**
Testing done: manual, uts

### CODE REVIEW

File: src/Agent.Worker/TestResults/ITestRunPublisher.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    public interface ITestRunPublisher
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/ITestRunPublisher.cs]: Change this to ServiceLocator Pattern. You do not have multiple implementation of this interface.
You could add a default Locator and call Get/Create Service.

Comment (bryanmacfarlane) [inline src/Agent.Worker/TestResults/ITestRunPublisher.cs]: +1

Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/ITestRunPublisher.cs]: +1

File: src/Test/NuGet.Config [hunk 2]
```diff
@@ -6,6 +6,7 @@
     <add key="dotnet-core" value="https://www.myget.org/F/dotnet-core/api/v3/index.json" />
     <add key="dotnet-buildtools" value="https://www.myget.org/F/dotnet-buildtools/api/v3/index.json" />    
     <add key="aspnetcidev" value="https://www.myget.org/F/aspnetcidev/api/v3/index.json" />
-    <add key="api.nuget.org" value="https://api.nuget.org/v3/index.json" />      
+    <add key="api.nuget.org" value="https://api.nuget.org/v3/index.json" />
+    <add key="mstest" value="\\aajohn3\share\packages" />
```
Comment (AshwiniChalla) [inline src/Test/NuGet.Config]: Will change this, once the nuget package is released. Will merge only after this is changed.

File: src/Agent.Worker/TestResults/FileZipper.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    internal class FileZipper : IZipper
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/FileZipper.cs]: Same goes for IZipper, please add a ServierLocator, you can avoid MockZipper totally.
We do not have a pattern of creating mockable (or testable) class created. The problem is we create testclasses and implement its functionality. But another test would require the testclass to behave in a different way, then we start customizing these testclasses. All these can go away with if you move to servicelocatoer.

In every test which requires your interface implementation, create a mock and setup your required methods and register with testhostcontext.

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+                reader = new TrxResultReader(executionContext);
+            }
+            else if (testRunner.Equals("JUnit", StringComparison.OrdinalIgnoreCase))
+            {
+                reader = new JUnitResultReader(executionContext);
+            }
+            else if (testRunner.Equals("NUnit", StringComparison.OrdinalIgnoreCase))
+            {
+                reader = new NUnitResultReader(executionContext);
+            }
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: You can use IExtensionManager to avoid this.
You could simply call in oneliner

`extensionManager.GetExtension<IResultReader>()).FirstOrDefault(x => x.Type.Equals(testRunner);`

Take a look at the ReleaseJobExtension, we have multiple implementation of IArtifactProvider and pick the right implementation by the type and avoid this if-else-if blocks.

File: src/Agent.Worker/TestResults/JunitResultReader.cs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    public class JUnitResultReader : IResultReader
+    {
+        public JUnitResultReader(IExecutionContext executionContext)
+        {
+            _executionContext = executionContext;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: There isn't any logs, you will want some logs for debugging. :)
You will need HostContext to trace

File: src/Agent.Worker/TestResults/JunitResultReader.cs [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            {
+                //first testsuite starteddate is the starteddate of the run
+                runSummary.TimeStamp = DateTime.Parse(runSummary.Results[0].StartedDate);
+            }
+
+            //create test run data
+            var testRunData = new TestRunData(
+                name: runSummary.Name,
+                startedDate: runSummary.TimeStamp.ToString("o"),
+                completedDate: runSummary.TimeStamp.Add(runSummary.Duration).ToString("o"),
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: Consider: looks like indentation is messed up and few empty lines below.

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    {
+        public Type ExtensionType
+        {
+            get
+            {
+                return typeof(ICommandExtension);
+            }
+        }
+
+        public String CommandArea
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: Consider: We generally use built-in type for primitive types. (string instead of String. int instead of Int32)

File: src/Agent.Worker/TestResults/TestRunPublisher.cs [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+                _executionContext.Warning(StringUtil.Loc("UnableToArchiveResults", ex));
+                UploadTestRunAttachmentsIndividual();
+            }
+        }
+
+        private void UploadTestRunAttachmentsIndividual()
+        {
+            _executionContext.Debug("Uploading test run attachements individually");
+            // Do not upload duplicate entries 
+            var attachedFiles = UniqueTestRunFiles;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/TestRunPublisher.cs]: Another style note: we use var only on the right hand side if the type is clear. If its not we explicitly define the type.

File: src/Agent.Worker/TestResults/TestRunPublisher.cs [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+                if (consoleLog.Length <= TCM_MAX_FILESIZE)
+                {
+                    byte[] bytes = System.Text.Encoding.UTF8.GetBytes(consoleLog);
+                    string encodedData = Convert.ToBase64String(bytes);
+                    return new TestAttachmentRequestModel(encodedData, consoleLogFileName, "",
+                        AttachmentType.ConsoleLog.ToString());
+                }
+                else
+                {
+                    StringUtil.Loc("AttachmentExceededMaximum", consoleLogFileName);
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/TestRunPublisher.cs]: Did you mean to throw an exception here?
StringUtil.Loc will only return a localized string, it has to be used by the caller.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+            string runName = Name + " Test Run";
+            if (runContext != null)
+            {
+                if (!String.IsNullOrWhiteSpace(runContext.RunName))
+                {
+                    runName = runContext.RunName;
+                }
+                else
+                {
+                    runName = string.Format("{0} {1} {2}", runName, runContext.Configuration, runContext.Platform);
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: stringUtil.Format

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+                        results.Add(resultCreateModel);
+                    }
+
+                }
+            });
+
+            return results;
+        }
+
+        private string _attachmentLocation;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: variables at the top

File: src/Agent.Worker/TestResults/XunitResultReader.cs [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+                    var runDate = assemblyNode.Attributes["run-date"].Value;
+                    var runTime = assemblyNode.Attributes["run-time"].Value;
+
+                    var startDate = DateTime.Now;
+                    var startTime = TimeSpan.Zero;
+                    if (DateTime.TryParse(runDate, out startDate))
+                    {
+                        if (TimeSpan.TryParse(runTime, out startTime))
+                        {
+                            // all good
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/XunitResultReader.cs]: add a log here? Otherwise you don't need the if condition at all here.

File: src/Test/_project.json [hunk 13]
```diff
@@ -37,20 +37,37 @@
...
-        "Agent.Listener": { "target": "project", "version": "1.0.0-*" },
+        "Agent.Listener": {
+            "target": "project",
+            "version": "1.0.0-*"
+        },
         "vss-api-netcore": "0.5.6-private",
-        "Microsoft.VisualStudio.Services.Agent": { "target": "project", "version": "1.0.0-*" },
-        "Agent.Worker": { "target": "project", "version": "1.0.0-*" } 
+        "Microsoft.VisualStudio.Services.Agent": {
+            "target": "project",
```
Comment (kasubram) [inline src/Test/_project.json]: IMO, the one line for dependency looked better.. I could scan the dependency on the left hand side in a single parse. Now its little bit hard to see what dependencies are required. @bryanmacfarlane @ericsciple any comments?

File: src/Misc/layoutbin/en-US/strings.json [hunk 14]
```diff
@@ -170,6 +176,7 @@
   "Prompt0Default1": "Enter {0} (press enter for {1})",
   "RegisterAgentSectionHeader": "Register Agent",
   "Replace": "replace? (Y/N)",
+  "ResultsCommandNotFound": "##vso[results.{0}] is not a recognized command for Task command extension. TODO: DOC aka link",
```
Comment (kasubram) [inline src/Misc/layoutbin/en-US/strings.json]: I prefixed all RM extension related string with "RM", May be its irrelevant. But it will be easier to identity string which were added for RM. Incase if I ever wanted to move the string to different place, it will be useful. Also we sort the string alphabetically, hence all RM strings are sorted out together. I don't know if everybody will agree to this. It was my choice.

File: src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+using System.Xml;
+using Microsoft.VisualStudio.Services.Agent.Worker.TestResults;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.TestResults
+{
+    [TestClass]
```
Comment (kasubram) [inline src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs]: We have been using Xunit for unit tests. Please take a look at the current tests.

File: src/Test/L0/Worker/TestResults/Mocks/MockTestManagementHttpClient.cs [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Worker.TestResults;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.TestResults
+{
+    public class MockTestManagementHttpClient : ITestManagementHttpClient
```
Comment (kasubram) [inline src/Test/L0/Worker/TestResults/Mocks/MockTestManagementHttpClient.cs]: I guess this is some testable class, this has lots of notimplementedexception. If you have a ServiceLocator, this class can go away and you can setup which ever method is required in your corresponding tests

File: src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs (hunk 15 above)
Comment (AshwiniChalla) [inline src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs]: We are onboarding mstest with this PR.

File: src/Agent.Worker/TestResults/FileZipper.cs (hunk 3 above)
Comment (bryanmacfarlane) [inline src/Agent.Worker/TestResults/FileZipper.cs]: Why is file zipper under the test results code.  Seems like it should be in common utilities and made generic

**Comment by bryanmacfarlane** [comment]
This needs many L0 tests covering positive and negative cases.  Testability and actually tested are two big initiatives with the new agent.

When this is merged in, it should be squashed.  There's a bunch of commits.

File: src/Agent.Worker/TestResults/FileZipper.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    [ServiceLocator(Default = typeof(FileZipper))]
+    public interface IZipper : IAgentService
+    {
+        void CreateZipFile(string zipfileName, IEnumerable<string> files);
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/FileZipper.cs]: There is only one place using this method, not sure whether worth to have it in a separate class other than a private method in the caller class.

File: src/Agent.Worker/TestResults/JunitResultReader.cs [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    public class JUnitResultReader : AgentService, IResultReader
+    {
+        public Type ExtensionType => typeof(IResultReader);
+
+        public JUnitResultReader()
+        {
+            AddResultsFileToRunLevelAttachments = true;
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: you don't really need the constructor. since you don't change it to anything.
public bool AddResultsFileToRunLevelAttachments {get {return true}}; already enought,

File: src/Agent.Worker/TestResults/JunitResultReader.cs [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+            if (TimeSpan.Compare(totalTestSuiteDuration, totalTestCaseDuration) < 0)
+            {
+                totalTestSuiteDuration = totalTestCaseDuration; //run duration may not be set in the xml, so use total test case duration 
+            }
+            testSuiteSummary.Duration = totalTestSuiteDuration;
+
+            return testSuiteSummary;
+        }
+
+        private static TimeSpan GetTimeSpan(XmlNode rootNode)
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: need static?

**Comment by TingluoHuang** [comment]
please follow CoreCLR style guidelines https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md
clean up all variable naming, code organize, etc.

**Comment by TingluoHuang** [comment]
better following the existing code pattern, use service locator for interface need testing. instead of create lots of mock, just using moq.

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+
+                    // publish test results to VSO using the reader based on the test runner specified
+                    var resultsReader = GetTestResultReader(testRunner);
+                    resultsReader.AddResultsFileToRunLevelAttachments = publishRunLevelAttachments;
+                    PublishTestResults(resultsReader);
+                }
+                catch (Exception ex)
+                {
+                    //Do not fail the task.
+                    string message = ex.InnerException == null ? ex.Message : ex.InnerException.Message;
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: why don't log the entire exception?

File: src/Agent.Worker/WorkerUtilties.cs [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+using System.Linq;
+using System.Text;
+using Microsoft.VisualStudio.Services.Client;
+using Microsoft.VisualStudio.Services.Common;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker
+{
+    public class WorkerUtilies
+    {
+        public static VssConnection GetVssConnection(IExecutionContext context)
```
Comment (TingluoHuang) [inline src/Agent.Worker/WorkerUtilties.cs]: we already have something in library, isn't it?

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+                //test case results
+                TestRunData resultFileRunData = publisher.ReadResultsFromFile(resultFile);
+
+                if (resultFileRunData != null && resultFileRunData.Results != null && resultFileRunData.Results.Length > 0)
+                {
+                    foreach (TestCaseResultData tcResult in resultFileRunData.Results)
+                    {
+                        int durationInMs = 0;
+                        Int32.TryParse(tcResult.DurationInMs, out durationInMs);
+                        totalTestCaseDuration = totalTestCaseDuration.Add(new TimeSpan(0, 0, 0, 0, durationInMs));
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: timespan.frommillisecond()

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 23]
```diff
@@ -0,0 +1,303 @@
...
+            eventProperties.TryGetValue(PublishTestResultsEventProperties.PublishRunAttachments, out publishRunAttachmentsInput);
+            if (string.IsNullOrEmpty(publishRunAttachmentsInput) || !bool.TryParse(publishRunAttachmentsInput, out publishRunLevelAttachments))
+            {
+                // if no proper input is provided by default we publish attachments.
+                publishRunLevelAttachments = true;
+            }
+        }
+
+        // for testing only
+        public List<string> warnings = new List<string>();
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: production code should not contains code that to support testing, if possible.

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 24]
```diff
@@ -0,0 +1,303 @@
...
+        // for testing this is made public.
+        public IExecutionContext executionContext;
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: same as this.

File: src/Agent.Worker/TestResults/TestRunPublisher.cs [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+
+        #region Public API
+        public void InitializePublisher(IExecutionContext executionContext, VssConnection connection, string projectName, TestRunContext runContext, IResultReader resultReader)
+        {
+            _executionContext = executionContext;
+            _projectName = projectName;
+            _runContext = runContext;
+            _resultReader = resultReader;
+            connection.InnerHandler.Settings.SendTimeout = TimeSpan.FromSeconds(PUBLISH_TIMEOUT);
+            _httpClient = connection.GetClient<TestManagementHttpClient>();
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TestRunPublisher.cs]: better follow the pattern Build and RM had, move all client behind interface, reference to BuildServer.cs and ReleaseServer.cs code.

File: src/Agent.Worker/TestResults/TestRunPublisher.cs [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+                    {
+                        Hashtable attachedFiles = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+                        foreach (string attachment in attachments)
+                        {
+                            if (!attachedFiles.ContainsKey(attachment))
+                            {
+                                TestAttachmentRequestModel reqModel = GetAttachmentRequestModel(attachment);
+                                if (reqModel != null)
+                                {
+                                    Task<TestAttachmentReference> trTask = _httpClient.CreateTestResultAttachmentAsync(reqModel, _projectName, _testRun.Id, testresults.Result[j].Id);
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TestRunPublisher.cs]: we are not using Wait() in the entire core agent. everything is async/await pattern.

**Comment by TingluoHuang** [comment]
use async/await instead of task.wait().

**Comment by TingluoHuang** [comment]
you are not respect cancellationtoken at all. need be fixed before merge.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 27]
```diff
@@ -0,0 +1,431 @@
...
+                }
+
+                lock (sync)
+                {
+                    if (resultCreateModel.AutomatedTestId != null && _definitions.ContainsKey(resultCreateModel.AutomatedTestId))
+                    {
+                        TestCaseDefinition definition = _definitions[resultCreateModel.AutomatedTestId];
+                        if (definition != null)
+                        {
+                            if (definition.Storage != null) resultCreateModel.AutomatedTestStorage = definition.Storage;
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: you are not following coding style.
if() 
{
///
}

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 28]
```diff
@@ -0,0 +1,431 @@
...
+                    resultCreateModel.AutomatedTestId = resultNode.Attributes["testId"].Value;
+                }
+
+                string executionId = null;
+                if (resultNode.Attributes["executionId"] != null && resultNode.Attributes["executionId"].Value != null)
+                {
+                    executionId = resultNode.Attributes["executionId"].Value;
+                }
+
+                lock (sync)
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: i am not clear the way you use lock. which object you are trying to protect?

**Comment by TingluoHuang** [comment]
AgentService is majorly for Service locator interface. it provide lot's of convenience, ex. tracing, hostcontext. you are not using any of them for most class that inherent from AgentService, please take reference the rest of agent code, find how to use it.

File: src/NuGet.Config [hunk 29]
```diff
@@ -6,6 +6,7 @@
     <add key="restsdk" value="https://www.myget.org/F/nugetvssprivate/api/v3/index.json"/>
     <add key="dotnet-core" value="https://www.myget.org/F/dotnet-core/api/v3/index.json" />
     <add key="dotnet-buildtools" value="https://www.myget.org/F/dotnet-buildtools/api/v3/index.json" />    
-    <add key="api.nuget.org" value="https://api.nuget.org/v3/index.json" />    
+    <add key="api.nuget.org" value="https://api.nuget.org/v3/index.json" />   
+    <add key="mstest" value="\\aajohn3\share\packages" />
```
Comment (TingluoHuang) [inline src/NuGet.Config]: remove this before merge.

File: src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs (hunk 15 above)
Comment (TingluoHuang) [inline src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs]: is mstest going to work on xplat?

**Comment by TingluoHuang** [comment]
I don't like the current tests you have, it's anti the current L0 test pattern.
1. it is using mstest instead of xunit, will this working on xplat
2. we already using moq, and have good helper method.

**Comment by bryanmacfarlane** [comment]
Agree with @Ting.  Consistency is critical here.  Follow the xunit, moq and AgentService patterns in the rest of the code

File: src/Agent.Worker/TestResults/FileZipper.cs (hunk 17 above)
Comment (bryanmacfarlane) [inline src/Agent.Worker/TestResults/FileZipper.cs]: I think you go one of two ways - private function (not much code here) or move it to common util and create an archive class

File: src/Agent.Worker/WorkerUtilties.cs (hunk 21 above)
Comment (AshwiniChalla) [inline src/Agent.Worker/WorkerUtilties.cs]: there is nothing that takes executioncontext as input.

File: src/Agent.Worker/TestResults/TestRunPublisher.cs (hunk 26 above)
Comment (AshwiniChalla) [inline src/Agent.Worker/TestResults/TestRunPublisher.cs]: We have 3 sequential steps here . start run, create attachments, end run. I will change end run to return as aync. other steps are sequential and cannot be synchronlized.

File: src/Agent.Worker/TestResults/TrxResultReader.cs (hunk 28 above)
Comment (AshwiniChalla) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: there is a parallel for each above .

**Comment by AshwiniChalla** [comment]
I have resolved the comments. Removed unit tests from PR. Will make that a different PR.

**Comment by bryanmacfarlane** [comment]
I would like big features like this to come with unit tests

**Comment by AshwiniChalla** [comment]
I agree with that Bryan. This is a long going PR now. We were struck on the choice of test framework xunit and mstest. Will make sure from next time code and uts go together.

**Comment by bryanmacfarlane** [comment]
The choice of test framework was already made.  We don't want next time.  We should close on it now.

File: src/Agent.Worker/WorkerUtilties.cs (hunk 21 above)
Comment (TingluoHuang) [inline src/Agent.Worker/WorkerUtilties.cs]: I see, maybe move this to IExecutionContext extension?

**Comment by bryanmacfarlane** [comment]
Aiming for new preview release Monday.  Please add tests back and get in to be in the preview.  I can wait another day if needed.

You should just be able to find replace the attributes and assert functions to make it work:  https://xunit.github.io/docs/comparisons.html 

Hope that helps.

File: src/Agent.Worker/TestResults/JunitResultReader.cs [hunk 30]
```diff
@@ -0,0 +1,313 @@
...
+
+        /// <summary>
+        /// Reads a JUnit results file from disk, converts it into a TestRunData object.        
+        /// </summary>
+        /// <param name="filePath"></param>
+        /// <param name="runContext"></param>
+        /// <returns></returns>
+        public TestRunData ReadResults(IExecutionContext executionContext, string filePath, TestRunContext runContext = null)
+        {
+            _executionContext = executionContext;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: You don't need this private variable. Its not used anywhere else. Keep it as local variables.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 31]
```diff
@@ -0,0 +1,448 @@
...
+            _executionContext.Debug(string.Format(CultureInfo.InvariantCulture, "Total test results: {0}", testRunData.Results.Length));
+
+            return testRunData;
+        }
+
+        public bool AddResultsFileToRunLevelAttachments
+        {
+            get
+            {
+                return _addResultsFileToRunLevelAttachments;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: You could just use the property itself right? why do you need another private variable? Its not a complex object also, its just bool.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 32]
```diff
@@ -0,0 +1,448 @@
...
+    internal class TestCaseDefinition
+    {
+        public string AutomatedTestName { get; private set; }
+        public string Storage { get; private set; }
+        public string Priority { get; private set; }
+        public IdentityRef Owner { get; private set; }
+
+        public TestCaseDefinition(string automatedTestName, IdentityRef owner, string priority, string storage)
+        {
+            this.AutomatedTestName = automatedTestName;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: coding style: avoid using "this"

File: src/Agent.Worker/TestResults/XunitResultReader.cs [hunk 33]
```diff
@@ -0,0 +1,227 @@
...
+            testRunData.Attachments = AddResultsFileToRunLevelAttachments ? new string[] { filePath } : new string[0];
+
+            return testRunData;
+        }
+
+        public bool AddResultsFileToRunLevelAttachments
+        {
+            get
+            {
+                return _addResultsFileToRunLevelAttachments;
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/XunitResultReader.cs]: same goes here.. why not use the property itself?

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 34]
```diff
@@ -0,0 +1,300 @@
...
+                publisher.StartTestRun(testRunData).Wait();
+                publisher.AddResults(runResults.ToArray()).Wait();
+                await publisher.EndTestRun(publishAttachmentsAsArchive: true);
+            }
+        }
+
+        /// <summary>
+        /// Publish separate test run for each result file that has results.
+        /// </summary>
+        private async Task PublishToNewTestRunPerTestResultFile(List<string> resultFiles, ITestRunPublisher publisher)
```
Comment (kasubram) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: coding style: async method should have method name ending with async.

Comment (AshwiniReddyChalla) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: will change

File: src/Agent.Worker/TestResults/XunitResultReader.cs (hunk 33 above)
Comment (AshwiniReddyChalla) [inline src/Agent.Worker/TestResults/XunitResultReader.cs]: correct. will change

File: src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs [hunk 35]
```diff
@@ -0,0 +1,307 @@
...
+        public void Dispose()
+        {
+            _jUnitReader.AddResultsFileToRunLevelAttachments = true;
+            try
+            {
+                File.Delete(_fileName);
+            }
+            catch
+            {
+
```
Comment (kasubram) [inline src/Test/L0/Worker/TestResults/JunitResultReaderTests.cs]: log it.

File: src/Test/L0/Worker/TestResults/NUnitResultReaderTests.cs [hunk 36]
```diff
@@ -0,0 +1,320 @@
...
+        public void Dispose()
+        {
+            _nUnitReader.AddResultsFileToRunLevelAttachments = true;
+            try
+            {
+                File.Delete(_fileName);
+            }
+            catch
+            {
+            }
```
Comment (kasubram) [inline src/Test/L0/Worker/TestResults/NUnitResultReaderTests.cs]: Add a trace?

File: src/Test/L0/Worker/TestResults/XUnitResultReaderTests.cs [hunk 37]
```diff
@@ -0,0 +1,400 @@
...
+            WriteXUnitFile(xunitResults);
+            XUnitResultReader reader = new XUnitResultReader();
+            TestRunData runData = reader.ReadResults(_ec.Object, _xUnitResultFile, new TestRunContext("Owner", "any cpu", "debug", 1, "", "releaseUri", "releaseEnvironmentUri"));
+
+            Assert.NotNull(runData);
+            Assert.Equal(0, runData.Results.Length);
+        }
+
+        [Fact]
+        [Trait("Level", "L0")]
```
Comment (kasubram) [inline src/Test/L0/Worker/TestResults/XUnitResultReaderTests.cs]: Add another Trait to add category
[Trait("Category", "PublishTests")] or something better.. :)

**Comment by kasubram** [comment]
Changes looks good from my side. Please address the comments before merging.

File: src/Agent.Worker/TestResults/ResultsCommands.cs [hunk 38]
```diff
@@ -0,0 +1,300 @@
...
+                    resultsReader.AddResultsFileToRunLevelAttachments = _publishRunLevelAttachments;
+                    PublishTestResults(resultsReader);
+                }
+                catch (Exception ex)
+                {
+                    //Do not fail the task.
+                    string message = ex.Message;
+                    if (ex.InnerException != null)
+                    {
+                        message += "\n";
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/ResultsCommands.cs]: NewLine Char is different in different platform, use Environment.NewLine instead.

File: src/Agent.Worker/TestResults/JunitResultReader.cs [hunk 39]
```diff
@@ -0,0 +1,308 @@
...
+
+            XmlDocument doc = new XmlDocument();
+            try
+            {
+                var settings = new XmlReaderSettings
+                {
+                    DtdProcessing = DtdProcessing.Prohibit
+                };
+
+                using (XmlReader reader = XmlReader.Create(filePath, settings))
```
Comment (acesiddhu) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: does this work on Linux and mac?>

Comment (AshwiniChalla) [inline src/Agent.Worker/TestResults/JunitResultReader.cs]: Yes. verified.
